gnc/sub8_perception/src/sub8_pcl/sub8_gateway.py

```diff
@@ -0,0 +1,202 @@
+#input_image = cv2.imread("images.png")  # testing image
+
+############### IMAGE PROCESSING ###############
+
+class Vector(object):
+    def __init__(self, unit_vector, magnitude, point1, point2):
+        self.unit = unit_vector
+        self.magnitude = magnitude
+        self.point_set = (point1, point2)
+
+    def __repr__(self):
+        return "unit vector: " + str(self.unit) + "; magnitude: " + str(self.magnitude) + "; " + str(self.point_set)
+
+    def equals(self, other):
+        if (self.magnitude == other.magnitude) and (set(self.unit) == set(other.unit)):
+            return True
+        else:
+            return False
+
+    @staticmethod
+    def unit_dot(v1, v2):
+        return (v1.unit[0] * v2.unit[0]) + (v1.unit[1] * v2.unit[1])
+
+    @staticmethod
+    def is_ortho(v1, v2, tolerance):
+        product = Vector.unit_dot(v1,v2)
+        if abs(product) < tolerance:
+            return True
+        else:
+            return False
+
+    @staticmethod
+    def is_parallel(v1, v2, tolerance):
+        product = Vector.unit_dot(v1,v2)
+        if (1 - abs(product)) < tolerance:
+            return True
+        else:
+            return False
+
+def color_filter(image, target_hue=None, t_hue=None, target_saturation=None, t_sat=None, target_brightness=None, t_brt=None):
+    hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
+    masks = []
+    if target_hue:
+        hue_matrix = hsv_image[:,:,0]
+        mask = cv2.inRange(hue_matrix, target_hue-t_hue, target_hue+t_hue)
+        masks.append(mask)
+    if target_saturation:
+        saturation_matrix = hsv_image[:,:,1]
+        mask = cv2.inRange(saturation_matrix, target_saturation-t_sat, target_saturation+t_sat)
+        masks.append(mask)
+    if target_brightness:
+        brightness_matrix = hsv_image[:,:,2]
+        mask = cv2.inRange(brightness_matrix, target_brightness-t_brt, target_brightness+t_brt)
+        masks.append(mask)
+    for mask in masks:
+        image = cv2.bitwise_and(image, image, mask=mask)
+    return image
+
+def edge_map(image):
+    grayscale = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+    edges = cv2.Canny(grayscale, 10, 255)
+    return edges
+
+def find_centroids(corners):
+    radius = 100
+    maximum = corners.max()
+    corners = cv2.threshold(corners, int(maximum*0.5), maximum, cv2.THRESH_BINARY)[1]
+    corners = np.uint8(corners)
+    temp = cv2.connectedComponentsWithStats(corners)
+    return temp[3]
+
+def corner_map(image):
+    image = cv2.cornerHarris(image,5,11,0.04)
+    return image
+
+def find_shape(point_set, edges):
+    tolerance = 0.1  # acceptable variation
+    # create the initial set of vectors
+    vectors = []
+    point_set = [list(point) for point in point_set]
+    point_set2 = [item for item in point_set]  # create copy
+    for point1 in point_set:
+        for point2 in point_set2:
+            if ((point2[0] != point1[0]) and (point2[1] != point1[1])):  # verify that the points are not the same

```

Comment:
I think you'll be a lot happier doing np.allclose instead of elementwise equality checking